### ISSUE TITLE
Adding multi threading optimization

### ISSUE THREAD

**Issue by zay448345045** [FixAkaTheFix/Tower-Pixel-Dungeon#4]
Hello developers.I would like to ask if the game has considered adding multi threading optimization. There are many monsters placed on the SPD based sandbox dungeon, but it runs very slow. Upon checking the CPU usage of the phone, it was found that only one core is in use, while the other cores are not.

**Comment by zay448345045** [comment]
The tower defense dungeons is even more severe, with thousands of monsters and items falling on the interface, causing fps to drop to single digits and severe lag. For large modules, parallel optimization is required.

**Comment by FixAkaTheFix** [comment]
Hi,
**Short answer:**
Sorry, I don't think that adding multithread is possible, with most pd modders being starting spare-time programmers and introducing multithreading being very effort and time-consuming
**Long answer:**
ShPD is not a big game and it's engine does not usually need multi-threading
The case with sandbox is towers working slowly is that actorthread is overloaded - most of the game, it seems to me, already runs on a couple of threads with one of them being an actor thread. This would be really hard to make mobs split into several threads of working in parallel, while the whole game works on a "turn priority" system, which almost must include all mobs being in one row, or thread, so they could be iterated in order by their act priority.
Also note: some of this game was inspired on soul knight's horde mode (if you know this game), and even those professional progs who made that game couldn't handle the sheer amount of mobs and projectiles being both effectively working and being in several threads without INSANE lag
Overall, this is a problem a bit too complex for an average indie programmer to solve, but I am working on a new version and I'll consider removing 300-rat waves from it)

**Comment by FixAkaTheFix** [comment]
Also, wanted to ask, what is your device and how much enemies, approximately, needed for the game to start lagging, mildly and severely

**Comment by zay448345045** [comment]
> Also, wanted to ask, what is your device and how much enemies, approximately, needed for the game to start lagging, mildly and severely

Processor is snapdragon8+gen1.
When about a thousand mobs appear.
Tower dungeon doesn't have lag.
There is a dungeon called the Too Cruel Dungeon. Some of the challenge options can increase the number of mobs, but after opening it, the game lags more seriously.
I think it's fun to challenge a large number of monsters, and the visual effects are very exciting!


![Screenshot_2024-06-25-23-56-13-675_ua dreamfools toocruelpixeldungeon dew](https://github.com/FixAkaTheFix/Tower-Pixel-Dungeon/assets/22346235/7afce0cf-ac1b-4cc6-9cc8-94b0c2b3c97a)

https://github.com/FixAkaTheFix/Tower-Pixel-Dungeon/assets/22346235/df884d5e-9bb5-44be-b754-2bbef109a464

**Comment by FixAkaTheFix** [comment]
Hi again!

First, the approaching infinity/ more mobs with more rooms challenge is... well... It will lag even on pc's, whatever the core quality is. Trust me, I played it. (And even multicore won't save you from the sheer amount of enemies you will see)
Second, glad that Towers doesn't lag! To mention, 
**I am the dev of pixel towers only**,
the dev of cruel pd is a user named Juh, he has his github, like every pd mod has it's own dev, and we mostly don't communicate
About Cruel pd you have problems with, as far as I know, it has stopped being developed quite a time ago, and optimization was never a thing anyone cosidered adding, as those million-mob more-roomed dungeon challenges, like too cruel's approaching infinity, were almost a joke from the start, which is mentioned in the game

Good luck completing ApInf. You'll need it)
Tpd developer,
TheFix

**Comment by zay448345045** [comment]
> Also, wanted to ask, what is your device and how much enemies, approximately, needed for the game to start lagging, mildly and severely

How about introducing the transifex framework for multilingual translation of game mods? Because through this simplified framework, community enthusiasts can participate in translation. If not, developers who understand PR usually provide game files containing translated text directly, which is very dependent on fate.

**Comment by FixAkaTheFix** [comment]
First, it is already done, without transifex. If anybody wanted to translate it, they would just need to write to me, and this is done. JDSALing translated Pixel Towers for Chinese, for example.
.
Also transifex is not free. As I said, I am an indie programmer and though I am trying to make the game of highest quality possible, I won't be using any external networks as this is again, a bit complex, time-consuming, and not necessary for such small (relatively to the original ShPD even) projects. Also transifex in particular is not free.

**Comment by zay448345045** [comment]
> 嗨，又来了！
> 
> 首先，即将到来的无限/更多拥有更多房间的生物挑战是......井。。。无论核心质量如何，它甚至在 PC 上也会滞后。相信我，我玩过。（即使是多核也无法将您从您将看到的大量敌人中拯救出来）其次，很高兴 Towers 没有滞后！值得一提的是，**我只是像素塔的开发**者，残酷的 pd 的开发者是一个名叫 Juh 的用户，他有自己的 github，就像每个 pd mod 都有自己的开发一样，我们大多不交流关于 Cruel pd 你有问题，据我所知，它很久以前就停止了开发， 优化从来都不是任何人喜欢添加的东西，因为那些百万生物的更多房间的地牢挑战，比如 Too Cruel 的 Approaching Infinity，从一开始就几乎是一个笑话，这在游戏中被提及
> 
> 祝你好运完成 ApInf。你会需要它） Tpd 开发人员，TheFix

Thank you for your reply! It's like this. Transifex mainly promotes a paid plan, but there are also free plan. The provided framework can complete basic translation, which can be completed slowly, or it can be merged by community enthusiasts submitting PR. Also, I think adding a large map and more tower and mobs challenge options to the tower dungeon would enhance the fun of the game and make the visual effect more impactful.

**Comment by zay448345045** [comment]
> First, it is already done, without transifex. If anybody wanted to translate it, they would just need to write to me, and this is done. JDSALing translated Pixel Towers for Chinese, for example. . Also transifex is not free. As I said, I am an indie programmer and though I am trying to make the game of highest quality possible, I won't be using any external networks as this is again, a bit complex, time-consuming, and not necessary for such small (relatively to the original ShPD even) projects. Also transifex in particular is not free.

Thank you for your reply! It's like this. Transifex mainly promotes a paid plan, but there are also free plan. The provided framework can complete basic translation, which can be completed slowly, or it can be merged by community enthusiasts submitting PR. Also, I think adding a large map and more tower and mobs challenge options to the tower dungeon would enhance the fun of the game and make the visual effect more impactful.
